channels_redis/alt.py

```diff
@@ -0,0 +1,311 @@
+                "Consumers are encouraged to use the common pattern:\n"
+                f"   self.channel_layer.group_add({repr(group)}, self.channel_name)"
+            )
+        group_channel = self._get_group_channel_name(group)
+        if group_channel not in self.groups:
+            self.groups[group_channel] = set()
+        group_channels = self.groups[group_channel]
+        if channel not in group_channels:
+            group_channels.add(channel)
+        shard = self._get_shard(group_channel)
+        await shard.subscribe(group_channel)
+
+    async def group_discard(self, group, channel):
+        """
+        Removes the channel from a group.
+        """
+        group_channel = self._get_group_channel_name(group)
+        assert group_channel in self.groups
+        group_channels = self.groups[group_channel]
+        assert channel in group_channels
+        group_channels.remove(channel)
+        if len(group_channels) == 0:
+            del self.groups[group_channel]
+            shard = self._get_shard(group_channel)
+            await shard.unsubscribe(group_channel)
+
+    async def group_send(self, group, message):
+        """
+        Send the message to all subscribers of the group.
+        """
+        group_channel = self._get_group_channel_name(group)
+        shard = self._get_shard(group_channel)
+        await shard.publish(group_channel, message)
+
+
+def on_close_noop(sender, exc=None):
+    """
+    If you don't pass an `on_close` function to the `Receiver`, then it
+    defaults to one that closes the Receiver whenever the last subscriber
+    unsubscribes. That is not what we want; instead, we want the Receiver
+    to continue even if no one is subscribed, because soon someone *will*
+    subscribe and we want things to continue from there. Passing this
+    empty function solves it.
+    """
+    pass
+
+
+class RedisSingleShardConnection:
+    def __init__(self, host, channel_layer):
+        self.host = host
+        self.channel_layer = channel_layer
+        self._subscribed_to = set()
+        self._lock = None
+        self._pub_conn = None
+        self._sub_conn = None
+        self._receiver = None
+        self._receive_task = None
+        self._keepalive_task = None
+
+    async def publish(self, channel, message):
+        conn = await self._get_pub_conn()
+        await conn.publish(channel, msgpack.packb(message))
+
+    async def subscribe(self, channel):
+        if channel not in self._subscribed_to:
+            self._subscribed_to.add(channel)
+            conn = await self._get_sub_conn()
+            await conn.subscribe(self._receiver.channel(channel))
+
+    async def unsubscribe(self, channel):
+        if channel in self._subscribed_to:
+            self._subscribed_to.remove(channel)
+            conn = await self._get_sub_conn()
+            await conn.unsubscribe(channel)
+
+    async def _get_pub_conn(self):
```

Comment:
I think that instead of getting a connection each time, each consumer should get their own pubsub client class.
And also, I think that there is no need of keeping track of subscribed channels locally.

Have a redis.pubsub client initialized and then get the channels subscribed by doing client.channels

Reply:
> I think that instead of getting a connection each time, each consumer should get their own pubsub client class.

So there would be as many connections to Redis as there are consumers? I think that is excessive, and probably less efficient.

> Have a redis.pubsub client initialized and then get the channels subscribed by doing client.channels

We need to keep our own set of channels so that we can resubscribe if we lose connection to Redis. `aioredis` doesn't reconnect nor resubscribe on its own, so we have to do that ourselves.

Reply:
>So there would be as many connections to Redis as there are consumers? I think that is excessive, and probably less efficient.

there would be as many connections, but we don't need to keep track of a set of connected clients.
Besides, this is the way to go. You need to have concurrent clients connected to redis and then do centralized publishes. Otherwise, this is no different from the current channel layer - where we have one single connection and use redis as a "message queue". 
We need to be broadcasting messages instead.

Reply:
>We need to keep our own set of channels so that we can resubscribe if we lose connection to Redis. aioredis doesn't reconnect nor resubscribe on its own, so we have to do that ourselves.

When we have separate connections per consumer,
why not detect if we lost connection from the consumer and reconnect individually?
this way we don't need to keep track of the clients in a set.

Reply:
Efficient broadcast happens in a tree-structure. One connection to Redis *per consumer* is not a tree structure and will over-burden your Redis server. In-memory operations are always faster than network operations. You are essentially proposing that we replace in-memory operations with more network I/O. It's not a good idea. I'm closing this discussion. Please trust your voice has been heard. Next I want to hear from the maintainers of this repo.